pkg/ipamd/datastore/data_store.go

```diff
@@ -701,7 +731,16 @@ func (ds *DataStore) UnassignPodIPv4Address(ipamKey IPAMKey) (ip string, deviceN
 
 	eni, addr := ds.eniPool.FindAddressForSandbox(ipamKey)
 
-	if BackfillMissingCheckpointFromCRI && addr == nil {
+	if addr == nil {
+		// This `if` block should be removed when the CRI
+		// migration code is finally removed.  Leaving a
+		// compile dependency here to make that obvious :P
+		var _ = CheckpointMigrationPhase
+
+		// If the entry was discovered by querying CRI at
+		// restart rather than by observing an ADD operation
+		// directly, then we won't have captured the true
+		// networkname/ifname.
```

Comment:
Can we have unified key for in-memory data store and File ? If so we can avoid the above logic. Also AssignIPv4Address doesnt account for key that has been built during restore(i.e.,  const BackfillNetworkName = "_migrated-from-cri" and const BackfillNetworkIface = "unknown")

Reply:
> Can we have unified key for in-memory data store and File

We do: `IPAMKey` is used for both in-memory and file, as you can see in the call to `FindAddressForSandbox` 2 lines above this diff chunk.

I think the placeholder values are not relevant for `AssignIPv4Address`?  `AssignIPv4Address` is invoked for _new_ CNI invocations, and the placeholder values are only used for assignments that were already created as part of a previous ipamd (and learned through CRI).

Oh, if you're referring to the ["return existing assignment for existing IPAMKey"](https://github.com/aws/amazon-vpc-cni-k8s/blob/170d3c12adcad0d56cfe7ced0f8ebf9a6b177ceb/pkg/ipamd/datastore/data_store.go#L480-L484) logic in `AssignPodIPv4Address`, then yes.  We can extend that to also include this _migrated-from-cri fallback, or I would rather just delete those lines tbh.  There is no legitimate reason that that function should ever be called twice for the same IPAMKey (without an intervening DEL) - I just postponed this conversation in my earlier PR...

Reply:
By unified key, i meant, can we populate entire key into file and restore it back so that we don't have to use const BackfillNetworkName = "_migrated-from-cri" and const BackfillNetworkIface = "unknown ?

Reply:
We can't.  The only case where we ever use these placeholder values is the CRI->file upgrade case where we don't _know_ the correct values (because CRI doesn't record/expose it).

Reply:
Thanks Gus for the detailed note. If i understand your explanation, CRI api called during nodeInit doesn't seem to provide all the information required to form complete IPAM key. If this is the case, then CNI spec guideline that you have quoted above cannot be met using current CRI API. I would say operational correctness should win over spec conformance and we can revisit conformance once CRI API can provide everything required to form the key. If CNI_CONTAINERID is guaranteed to be unique enough, we could just have this as part of the key